lib/swaps/Swaps.ts

```diff
@@ -349,59 +349,55 @@ class Swaps extends EventEmitter {
-    assert(deal.state === SwapDealState.Normal, 'deal state is Error. Can not change deal status');
+    assert(deal.state === SwapDealState.Active, 'deal is not Active. Can not change deal status');
+
     switch (newStatus) {
       case SwapDealStatus.SwapCreated:
         assert(false, 'can not set deal status to SwapCreated.');
         break;
       case SwapDealStatus.SwapRequested:
         assert(deal.myRole === SwapDealRole.Taker, 'SwapRequested can only be set by the taker');
         assert(deal.status === SwapDealStatus.SwapCreated, 'SwapRequested can be only be set after SwapCreated');
+        this.logger.debug('Requesting deal: ' + JSON.stringify(deal));
         break;
       case SwapDealStatus.SwapAgreed:
         assert(deal.myRole === SwapDealRole.Maker, 'SwapAgreed can only be set by the maker');
         assert(deal.status === SwapDealStatus.SwapCreated, 'SwapAgreed can be only be set after SwapCreated');
+        this.logger.debug('Sending swap response to peer ');
         break;
       case SwapDealStatus.AmountSent:
         assert(deal.myRole === SwapDealRole.Taker && deal.status === SwapDealStatus.SwapRequested ||
           deal.myRole === SwapDealRole.Maker && deal.status === SwapDealStatus.SwapAgreed,
             'AmountSent can only be set after SwapRequested (taker) or SwapAgreed (maker)');
+        deal.executeTime = Date.now();
         break;
       case SwapDealStatus.AmountReceived:
         assert(deal.status === SwapDealStatus.AmountSent, 'AmountReceived can be only be set after AmountSent');
+        this.logger.debug('Amount received for preImage ' + deal.r_preimage);
         break;
       case SwapDealStatus.SwapCompleted:
         assert(deal.status === SwapDealStatus.AmountReceived, 'SwapCompleted can be only be set after AmountReceived');
+        deal.competionTime = Date.now();
+        this.setDealState(deal, SwapDealState.Completed, 'Swap completed. preimage = ' + deal.r_preimage);
+        this.logger.debug('Swap completed. preimage = ' + deal.r_preimage);
         break;
       default:
         assert(false, 'unknown deal status');
     }
 
     deal.status = newStatus;
-    switch (deal.status) {
-      case SwapDealStatus.SwapRequested:
-        this.logger.debug('Requesting deal: ' + JSON.stringify(deal));
-        break;
-      case SwapDealStatus.SwapAgreed:
-        this.logger.debug('Sending swap response to peer ');
-        break;
-      case SwapDealStatus.AmountSent:
-        deal.executeTime = Date.now();
-        break;
-      case SwapDealStatus.AmountReceived:
-        this.logger.debug('Amount received for preImage ' + deal.r_preimage);
-        this.emit('swap.paid', deal);
-        break;
-      case SwapDealStatus.SwapCompleted:
-        deal.competionTime = Date.now();
-        this.logger.debug('Swap completed. preimage = ' + deal.r_preimage);
-        break;
+
+    if (deal.status === SwapDealStatus.AmountReceived) {
+      this.emit('swap.paid', deal);
```

Comment:
This can also go in the `case SwapDealStatus.AmountReceived:` logic right, or is there a reason it's separate? Again my thinking is just to have all logic in one place and not make more checks than needed.

Reply:
A possible race between the listener and the code that change the status.
Maybe low risk today but if some code added in the future this may break.

Reply:
OK, if you want to leave it this way to clearly set the status before any listeners are informed that's fine. We *shouldn't* even be looking at the status from listeners - if we receive `swap.paid` we already know that it's been paid - and as you said this is supposed to be more of an internal indicator, but this way doesn't hurt. So I'd say either way is fine, up to you.